homeassistant/components/ping/binary_sensor.py

```diff
@@ -131,23 +133,50 @@ def __init__(self, host, count):
                 self._ip_address,
             ]
 
-    def ping(self):
+    async def async_ping(self):
         """Send ICMP echo request and return details if success."""
-        pinger = subprocess.Popen(
-            self._ping_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        pinger = await asyncio.create_subprocess_exec(
+            *self._ping_cmd,
+            stdin=None,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
         )
         try:
-            out = pinger.communicate(timeout=self._count + PING_TIMEOUT)
-            _LOGGER.debug("Output is %s", str(out))
+            out_data, out_error = await asyncio.wait_for(
+                pinger.communicate(), self._count + PING_TIMEOUT
+            )
+
+            if out_data:
+                _LOGGER.debug(
+                    "Output of command: `%s`, return code: %s:\n%s",
+                    " ".join(self._ping_cmd),
+                    pinger.returncode,
+                    out_data,
+                )
+            if out_error:
+                _LOGGER.debug(
+                    "Error of command: `%s`, return code: %s:\n%s",
+                    " ".join(self._ping_cmd),
+                    pinger.returncode,
+                    out_error,
+                )
+
+            if pinger.returncode != 0:
+                _LOGGER.exception(
+                    "Error running command: `%s`, return code: %s",
+                    " ".join(self._ping_cmd),
+                    pinger.returncode,
+                )
+
             if sys.platform == "win32":
-                match = WIN32_PING_MATCHER.search(str(out).split("\n")[-1])
+                match = WIN32_PING_MATCHER.search(str(out_data).split("\n")[-1])
                 rtt_min, rtt_avg, rtt_max = match.groups()
                 return {"min": rtt_min, "avg": rtt_avg, "max": rtt_max, "mdev": ""}
-            if "max/" not in str(out):
-                match = PING_MATCHER_BUSYBOX.search(str(out).split("\n")[-1])
+            if "max/" not in str(out_data):
+                match = PING_MATCHER_BUSYBOX.search(str(out_data).split("\n")[-1])
                 rtt_min, rtt_avg, rtt_max = match.groups()
                 return {"min": rtt_min, "avg": rtt_avg, "max": rtt_max, "mdev": ""}
-            match = PING_MATCHER.search(str(out).split("\n")[-1])
+            match = PING_MATCHER.search(str(out_data).split("\n")[-1])
             rtt_min, rtt_avg, rtt_max, rtt_mdev = match.groups()
             return {"min": rtt_min, "avg": rtt_avg, "max": rtt_max, "mdev": rtt_mdev}
         except subprocess.TimeoutExpired:
```

Comment:
```suggestion
        except asyncio.TimeoutError:
```

Reply:
Then you need to add something like
```
            _LOGGER.exception(
                "Timed out running command: `%s`, after: %ss", self._ping_cmd, self._count + PING_TIMEOUT
            )
            if pinger:
                try:
                    await pinger.kill()
                except TypeError:
                    pass
                del pinger

            return
```

Reply:
what about ProcessLookupError (like shown here: https://stackoverflow.com/a/45026089/965722)? should I also add except and pass for it?